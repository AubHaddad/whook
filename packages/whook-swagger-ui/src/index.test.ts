import { jest } from '@jest/globals';
import { constant, initializer } from 'knifecycle';
import { default as axios } from 'axios';
import {
  prepareServer,
  prepareEnvironment,
  initGetPingDefinition,
} from '@whook/whook';
import initHTTPRouter from '@whook/http-router';
import wrapHTTPRouterWithSwaggerUI from './index.js';
import { YError } from 'yerror';
import type { OpenAPIV3 } from 'openapi-types';
import type { Logger } from 'common-services';

describe('wrapHTTPRouterWithSwaggerUI', () => {
  const HOST = 'localhost';
  const PORT = 22222;
  const BASE_PATH = '/v1';
  const API: OpenAPIV3.Document = {
    openapi: '3.0.2',
    info: {
      version: '1.0.0',
      title: 'Sample OpenAPI',
      description: 'A sample OpenAPI file for testing purpose.',
    },
    paths: {
      [initGetPingDefinition.path]: {
        [initGetPingDefinition.method]: initGetPingDefinition.operation,
      },
    },
  };
  const logger = {
    output: jest.fn(),
    error: jest.fn(),
    debug: jest.fn(),
  };
  const time = jest.fn();
  const $autoload = jest.fn(async (serviceName) => {
    throw new YError('E_UNMATCHED_DEPENDENCY', serviceName);
  });
  let $;

  beforeEach(() => {
    logger.output.mockReset();
    logger.error.mockReset();
    logger.debug.mockReset();
    time.mockReset();
    $autoload.mockClear();
  });

  beforeEach(async () => {
    $ = await prepareEnvironment();

    $.register(
      initializer(
        {
          name: '$autoload',
          type: 'service',
          singleton: true,
        },
        async () => $autoload,
      ),
    );
    $.register(constant('BASE_PATH', BASE_PATH));
    $.register(constant('API', API));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEV_ACCESS_TOKEN', 'oudelali'));
    $.register(constant('HOST', HOST));
    $.register(constant('WRAPPERS', []));
    $.register(
      constant('HANDLERS', {
        getPing: jest.fn(async () => ({
          status: 200,
          headers: { 'content-type': 'application/json' },
          body: { ping: 'pong' },
        })),
      }),
    );
    $.register(constant('logger', logger as Logger));
    $.register(constant('time', time));
    $.register(constant('NODE_ENVS', ['test']));
  });

  it('should work', async () => {
    $.register(constant('PORT', PORT));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(constant('CONFIG', {}));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));
    $.register(
      constant('ENV', {
        DEV_MODE: '1',
      }),
    );

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios({
      method: 'get',
      url: `http://${HOST}:${PORT}${BASE_PATH}/ping`,
      headers: { 'user-agent': '__avoid_axios_version__' },
      validateStatus: () => true,
    });

    await $instance.destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: undefined,
        etag: undefined,
        'last-modified': undefined,
        server: undefined,
      },
      data,
      debugCalls: logger.debug.mock.calls.sort(sortLogs),
      logInfoCalls: logger.output.mock.calls.map(removeIps),
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should serve Swagger HTML', async () => {
    $.register(constant('PORT', PORT + 2));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(
      constant('CONFIG', {
        localURL: `http://${HOST}:${PORT + 2}`,
      }),
    );
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', ['test']));
    $.register(
      constant('ENV', {
        DEV_MODE: '1',
      }),
    );

    time.mockReturnValue(new Date('2010-03-06T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );
    const { status, headers, data } = await axios({
      method: 'get',
      url: `http://${HOST}:${PORT + 2}/docs`,
      headers: { 'user-agent': '__avoid_axios_version__' },
      validateStatus: () => true,
    });

    await $instance.destroy();

    expect({
      status,
      headers: {
        ...headers,
        // Erasing the Date header that may be added by Axios :/
        date: undefined,
        etag: undefined,
        'last-modified': undefined,
        server: undefined,
      },
      data,
      debugCalls: logger.debug.mock.calls.sort(sortLogs),
      logInfoCalls: logger.output.mock.calls.map(removeIps),
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });

  it('should be bypassed with no debug env', async () => {
    $.register(constant('PORT', PORT + 1));
    $.register(wrapHTTPRouterWithSwaggerUI(initHTTPRouter));
    $.register(constant('NODE_ENV', 'test'));
    $.register(constant('DEBUG_NODE_ENVS', []));
    $.register(constant('ENV', {}));

    time.mockReturnValue(new Date('2012-01-15T00:00:00Z').getTime());

    const { $instance } = await prepareServer(
      ['$instance', 'httpServer', 'process'],
      $,
    );

    await $instance.destroy();

    expect({
      debugCalls: logger.debug.mock.calls.sort(sortLogs),
      logInfoCalls: logger.output.mock.calls.map(removeIps),
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });
});

function sortLogs(strs1, strs2) {
  return strs1[0] > strs2[0] ? 1 : strs1[0] === strs2[0] ? 0 : -1;
}

function removeIps(strs) {
  return strs.map((str) =>
    (str as string).replace(/(127\.0\.0\.1|::1)/gm, '_ip_'),
  );
}
