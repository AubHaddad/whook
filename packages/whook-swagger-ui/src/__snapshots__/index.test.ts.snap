// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHTTPRouterWithSwaggerUI should be bypassed with no debug env 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "SOCKET_TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
    Array [
      "ERRORS_DESCRIPTORS",
    ],
    Array [
      "DEFAULT_ERROR_CODE",
    ],
    Array [
      "SHIELD_CHAR",
    ],
    Array [
      "MAX_CLEAR_CHARS",
    ],
    Array [
      "MAX_CLEAR_RATIO",
    ],
    Array [
      "SENSIBLE_PROPS",
    ],
    Array [
      "SENSIBLE_HEADERS",
    ],
  ],
  "debugCalls": Array [
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚ù§Ô∏è - Initializing the APM service.",
    ],
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üï∂Ô∏è - Initializing the obfuscator service.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "üõÇ - Initializing the importer!",
    ],
  ],
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:22223\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [],
}
`;

exports[`wrapHTTPRouterWithSwaggerUI should serve Swagger HTML 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "SOCKET_TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
    Array [
      "ERRORS_DESCRIPTORS",
    ],
    Array [
      "DEFAULT_ERROR_CODE",
    ],
    Array [
      "SHIELD_CHAR",
    ],
    Array [
      "MAX_CLEAR_CHARS",
    ],
    Array [
      "MAX_CLEAR_RATIO",
    ],
    Array [
      "SENSIBLE_PROPS",
    ],
    Array [
      "SENSIBLE_HEADERS",
    ],
  ],
  "data": "<!-- HTML for static distribution bundle build -->
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <title>Swagger UI</title>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"./swagger-ui.css\\" >
    <link rel=\\"icon\\" type=\\"image/png\\" href=\\"./favicon-32x32.png\\" sizes=\\"32x32\\" />
    <link rel=\\"icon\\" type=\\"image/png\\" href=\\"./favicon-16x16.png\\" sizes=\\"16x16\\" />
    <style>
      html
      {
        box-sizing: border-box;
        overflow: -moz-scrollbars-vertical;
        overflow-y: scroll;
      }

      *,
      *:before,
      *:after
      {
        box-sizing: inherit;
      }

      body
      {
        margin:0;
        background: #fafafa;
      }
    </style>
  </head>

  <body>
    <div id=\\"swagger-ui\\"></div>

    <script src=\\"./swagger-ui-bundle.js\\"> </script>
    <script src=\\"./swagger-ui-standalone-preset.js\\"> </script>
    <script>
    window.onload = function() {
      // Begin Swagger UI call region
      const ui = SwaggerUIBundle({
        url: \\"https://petstore.swagger.io/v2/swagger.json\\",
        dom_id: '#swagger-ui',
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: \\"StandaloneLayout\\"
      })
      // End Swagger UI call region

      window.ui = ui
    }
  </script>
  </body>
</html>
",
  "debugCalls": Array [
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚ù§Ô∏è - Initializing the APM service.",
    ],
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üï∂Ô∏è - Initializing the obfuscator service.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "üõÇ - Dynamic import of: ecstatic",
    ],
    Array [
      "üõÇ - Dynamic import of: swagger-ui-dist",
    ],
    Array [
      "üõÇ - Initializing the importer!",
    ],
  ],
  "headers": Object {
    "cache-control": "max-age=3600",
    "connection": "close",
    "content-length": "1424",
    "content-type": "text/html",
    "date": undefined,
    "etag": undefined,
    "last-modified": undefined,
    "server": undefined,
  },
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üíÅ - Serving the public API docs: http://localhost:22224/docs?url=http%3A%2F%2Flocalhost%3A22224%2Fv1%2FopenAPI",
    ],
    Array [
      "üíÅ - Serving the private API docs: http://localhost:22224/docs?url=http%3A%2F%2Flocalhost%3A22224%2Fv1%2FopenAPI%3Faccess_token%3Doudelali",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:22224\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [],
  "status": 200,
}
`;

exports[`wrapHTTPRouterWithSwaggerUI should work 1`] = `
Object {
  "autoloaderCalls": Array [
    Array [
      "MAX_HEADERS_COUNT",
    ],
    Array [
      "KEEP_ALIVE_TIMEOUT",
    ],
    Array [
      "SOCKET_TIMEOUT",
    ],
    Array [
      "MAX_CONNECTIONS",
    ],
    Array [
      "PROCESS_NAME",
    ],
    Array [
      "SIGNALS",
    ],
    Array [
      "BUFFER_LIMIT",
    ],
    Array [
      "PARSERS",
    ],
    Array [
      "STRINGIFYERS",
    ],
    Array [
      "DECODERS",
    ],
    Array [
      "ENCODERS",
    ],
    Array [
      "QUERY_PARSER",
    ],
    Array [
      "TIMEOUT",
    ],
    Array [
      "TRANSACTIONS",
    ],
    Array [
      "uniqueId",
    ],
    Array [
      "ERRORS_DESCRIPTORS",
    ],
    Array [
      "DEFAULT_ERROR_CODE",
    ],
    Array [
      "SHIELD_CHAR",
    ],
    Array [
      "MAX_CLEAR_CHARS",
    ],
    Array [
      "MAX_CLEAR_RATIO",
    ],
    Array [
      "SENSIBLE_PROPS",
    ],
    Array [
      "SENSIBLE_HEADERS",
    ],
  ],
  "data": Object {
    "ping": "pong",
  },
  "debugCalls": Array [
    Array [
      "‚åõ - Delay service initialized.",
    ],
    Array [
      "‚è≥ - Cancelling pending timeouts:",
      0,
    ],
    Array [
      "‚è≥ - Cleared a delay",
    ],
    Array [
      "‚è≥ - Created a delay:",
      30000,
    ],
    Array [
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "‚úîÔ∏è - HTTP server closed!",
    ],
    Array [
      "‚ù§Ô∏è - Initializing the APM service.",
    ],
    Array [
      "üë£ - Logging service initialized.",
    ],
    Array [
      "üí± - HTTP Transaction initialized.",
    ],
    Array [
      "üìá - Process service initialized.",
    ],
    Array [
      "üï∂Ô∏è - Initializing the obfuscator service.",
    ],
    Array [
      "üö¶ - HTTP Router initialized.",
    ],
    Array [
      "üõÇ - Dynamic import of: ecstatic",
    ],
    Array [
      "üõÇ - Dynamic import of: swagger-ui-dist",
    ],
    Array [
      "üõÇ - Initializing the importer!",
    ],
  ],
  "headers": Object {
    "connection": "close",
    "content-type": "application/json",
    "date": undefined,
    "etag": undefined,
    "last-modified": undefined,
    "server": undefined,
    "transaction-id": "0",
    "transfer-encoding": "chunked",
  },
  "logErrorCalls": Array [
    Array [
      "üîÇ - Running in \\"test\\" environment.",
    ],
    Array [
      "üíÅ - Serving the public API docs: http://localhost:22222/docs?url=http%3A%2F%2Flocalhost%3A22222%2Fv1%2FopenAPI",
    ],
    Array [
      "üíÅ - Serving the private API docs: http://localhost:22222/docs?url=http%3A%2F%2Flocalhost%3A22222%2Fv1%2FopenAPI%3Faccess_token%3Doudelali",
    ],
    Array [
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:22222\\".",
    ],
    Array [
      "On air üöÄüåï",
    ],
  ],
  "logInfoCalls": Array [
    Array [
      "CALL",
      "{\\"id\\":\\"0\\",\\"protocol\\":\\"http\\",\\"ip\\":\\"127.0.0.1\\",\\"startInBytes\\":147,\\"startOutBytes\\":0,\\"startTime\\":1267833600000,\\"url\\":\\"/v1/ping\\",\\"method\\":\\"GET\\",\\"reqHeaders\\":{\\"accept\\":\\"application/json, text/plain, */*\\",\\"user-agent\\":\\"__avoid_axios_version__\\",\\"host\\":\\"localhost:22222\\",\\"connection\\":\\"close\\"},\\"errored\\":false,\\"endTime\\":1267833600000,\\"endInBytes\\":147,\\"endOutBytes\\":179,\\"statusCode\\":200,\\"resHeaders\\":{\\"content-type\\":\\"application/json\\"},\\"operationId\\":\\"getPing\\"}",
    ],
  ],
  "status": 200,
}
`;
